### PR TITLE
OM-824 | Add custom order filter to order primary contacts

### PR DESCRIPTION
**Pull request by pjlampil**
Custom ordering filter to be able to order by primary contacts. Originally the need was to order by primary address city, but now the solution is generic enough to be able to order by any `Address` field oor by primary `Email`

### CODE REVIEW

File: profiles/schema.py [hunk 1]
```diff
@@ -172,6 +175,62 @@ def resolve_total_count(self, info, **kwargs):
...
+        ]
+
+    def filter(self, qs, values):
+        def _get_field_data_by_field(fields, field):
+            for f in fields:
+                if f[0] == field:
+                    return f
+            return None
+
+        for v in values or []:
```
Comment (igordavydsson) [inline profiles/schema.py:213]: isn't `values` empty list by default? 🤔

Comment (pjlampil) [reply profiles/schema.py:213]: Nope, it's `None` 😭

Comment (igordavydsson) [reply profiles/schema.py]: :/
alright then

File: profiles/schema.py [hunk 2]
```diff
@@ -172,6 +176,65 @@ def resolve_total_count(self, info, **kwargs):
...
+        ]
+
+    def filter(self, qs, values):
+        def _get_field_data_by_field(fields, field):
+            for f in fields:
+                if f[0] == field:
+                    return f
+            return None
+
+        def _camel_to_snake(value):
```
Comment (igordavydsson) [inline profiles/schema.py:214]:
```python
from graphene.utils.str_converters import to_snake_case
```
☝️ 🤔 ?

Comment (pjlampil) [reply profiles/schema.py:214]: Sure thing. Didn't know this existed.

File: profiles/schema.py [hunk 3]
```diff
@@ -172,6 +176,62 @@ def resolve_total_count(self, info, **kwargs):
...
+        for value in values or []:
+            # match with all of our custom ascending and descending orderings
+            if value in options:
+                # get rid of leading "-" if ordering is descending
+                field_name = value.replace("-", "")
+                field_data = _get_field_data_by_field(self.FIELDS, field_name)
+                # annotate the query with extra field and return the queryset
+                annotation = {
+                    to_snake_case(field_name): Subquery(
+                        field_data[2]
```
Comment (igordavydsson) [inline profiles/schema.py:226]: so, if `field_data == None` you get nice exceptions here? 🙂

Comment (pjlampil) [reply profiles/schema.py:226]: Yep. I was just iterating the idea here hence the draft pr. I'm still going to make some checks there and actually tests alone will keep malicious configurations out of the production.

Comment (igordavydsson) [reply profiles/schema.py:226]: but yeah, it should not get here ever, since you check `if value in options`...

Comment (pjlampil) [reply profiles/schema.py:226]: Yeah. I guess I'm still going to do the absolutely necessary checks to the final pr but the tests alone will catch most of the toxic configuration 😃

File: profiles/schema.py [hunk 4]
```diff
@@ -172,6 +176,62 @@ def resolve_total_count(self, info, **kwargs):
...
+        self.extra["choices"] += [
+            (f"-{item[0]}", f"{item[1]} (Descending)") for item in self.FIELDS
+        ]
+
+    def filter(self, qs, values):
+        def _get_field_data_by_field(fields, field):
+            for f in fields:
+                if f[0] == field:
+                    return f
+            return None
```
Comment (igordavydsson) [inline profiles/schema.py:212]: you could probably avoid some loops and chain if you turn `FIELDS` into a dict:
```python
FIELDS = {
    "primaryCity": {"label": "Primary City", "model": Address, "field_name": "city"},
    "primaryPostalCode": {"label": "Primary Postal Code", "model": Address, "field_name": "postal_code"},
    "primaryAddress": {"label": "Primary Address", "model": Address, "field_name": "address"},
    "primaryCountryCode": {"label": "Primary Country Code", "model": Address, "field_name": "country_code"},
    "primaryEmail": {"label": "Primary Email", "model": Email, "field_name": "email"},
}

def __init__(self, *args, **kwargs):
    super().__init__(*args, **kwargs)
    # add ascending and descending custom orderings
    self.custom_ordering_options = []
    for field, field_data in self.FIELDS.items():
        self.extra["choices"] += [(field, field_data["label"]), (f"-{field}", f"{field_data["label"]} (Descending)")
        self.custom_ordering_options.extend([field, f"-{field}"])
```
and then you could do `if value in self.custom_ordering_options` or somehting 🤔 
not sure if this looks any better or actually adds readability, you are free to keep your implementation or take some parts from this sugegstions

File: requirements.in [hunk 5]
```diff
@@ -16,7 +16,7 @@ django-parler
 django-reversion
 django-searchable-encrypted-fields
 django-thesaurus>=0.0.2
-graphene-django
+graphene-django>=2.10.1
```
Comment (igordavydsson) [inline requirements.in:19]: could have maybe just updated all the deps 🙂   not necessary, just a suggestion for future situations like this

Comment (pjlampil) [reply requirements.in:20]: Ayeaye. I was deep in a hole when I did that version bumb (which actually solved all my problems). Good suggestion 🎖️

**Pull request merged**